rust/noosphere/src/ffi/fs.rs

```diff
@@ -0,0 +1,233 @@
+use anyhow::anyhow;
+use noosphere_core::data::Did;
+use noosphere_fs::{SphereFile, SphereFs};
+use safer_ffi::prelude::*;
+use std::{pin::Pin, str::FromStr};
+use subtext::{Peer, Slashlink};
+use tokio::io::{AsyncRead, AsyncReadExt};
+
+use crate::{
+    ffi::{NsHeaders, NsNoosphereContext},
+    platform::{PlatformKeyMaterial, PlatformStore},
+};
+
+ReprC! {
+    #[ReprC::opaque]
+    pub struct NsSphereFs {
+        inner: SphereFs<PlatformStore, PlatformKeyMaterial>
+    }
+}
+
+impl NsSphereFs {
+    pub fn inner(&self) -> &SphereFs<PlatformStore, PlatformKeyMaterial> {
+        &self.inner
+    }
+
+    pub fn inner_mut(&mut self) -> &mut SphereFs<PlatformStore, PlatformKeyMaterial> {
+        &mut self.inner
+    }
+}
+
+ReprC! {
+    #[ReprC::opaque]
+    pub struct NsSphereFile {
+        inner: SphereFile<Pin<Box<dyn AsyncRead>>>
+    }
+}
+
+impl NsSphereFile {
+    pub fn inner(&self) -> &SphereFile<Pin<Box<dyn AsyncRead>>> {
+        &self.inner
+    }
+
+    pub fn inner_mut(&mut self) -> &mut SphereFile<Pin<Box<dyn AsyncRead>>> {
+        &mut self.inner
+    }
+}
+
+#[ffi_export]
+/// Initialize an instance of a [NsSphereFs] that is a read/write view into
+/// the contents (as addressed by the slug namespace) of the identitifed sphere
+pub fn ns_sphere_fs_open(
+    noosphere: &NsNoosphereContext,
+    sphere_identity: char_p::Ref<'_>,
+) -> repr_c::Box<NsSphereFs> {
+    noosphere
+        .async_runtime()
+        .block_on(async {
+            let sphere_context = noosphere
+                .inner()
+                .get_sphere_context(&Did(sphere_identity.to_str().into()))
+                .await?;
+
+            let sphere_context = sphere_context.lock().await;
+
+            Ok(repr_c::Box::new(NsSphereFs {
+                inner: sphere_context.fs().await?,
+            })) as Result<_, anyhow::Error>
+        })
+        .unwrap()
+}
+
+#[ffi_export]
+/// De-allocate an [NsSphereFs] instance
+pub fn ns_sphere_fs_free(sphere_fs: repr_c::Box<NsSphereFs>) {
```

Comment:
To be clear, I think `open`/`free` pairs work fine here, just thinking through how we use `open` vs `create`

Reply:
`free` seems like useful jargon because it communicates: "once you have done this, the pointer is no longer able to be used."

In all cases so far, when you get a pointer to something, you have to manually free it. Resources with long-lived lifetimes separate from those pointers (like `SphereContext`) are managed internally by library.

I agree that the language isn't ideal. I'm happy to conventionalize any specific language, but short of a concrete proposal I'll probably keep this as-is.